scriptworker/github.py

```diff
@@ -65,3 +76,152 @@ def get_release(self, tag_name):
+        return html_text != ''
+
+
+def is_github_url(url):
+    """Tell if a given URL matches a Github one.
+
+    Args:
+        url (str): The URL to test. It can be None.
+
+    Returns:
+        bool: False if the URL is not a string or if it doesn't match a Github URL
+
+    """
+    if isinstance(url, str):
+        return url.startswith('https://github.com/')
+    else:
+        return False
+
+
+def extract_github_repo_owner_and_name(url):
+    """Given an URL, return the repo name and who owns it.
+
+    Args:
+        url (str): The URL to the GitHub repository
+
+    Raises:
+        ValueError: on url that aren't from github
+
+    Returns:
+        str, str: the owner of the repository, the repository name
+
+    """
+    _check_github_url_is_supported(url)
+
+    parts = get_parts_of_url_path(url)
+    repo_owner = parts[0]
+    repo_name = parts[1]
+
+    return repo_owner, _strip_trailing_dot_git(repo_name)
+
+
+def extract_github_repo_and_revision_from_source_url(url):
+    """Given an URL, return the repo name and who owns it.
+
+    Args:
+        url (str): The URL to the GitHub repository
+
+    Raises:
+        ValueError: on url that aren't from github or when the revision cannot be extracted
+
+    Returns:
+        str, str: the owner of the repository, the repository name
+
+    """
+    _check_github_url_is_supported(url)
+
+    parts = get_parts_of_url_path(url)
+    repo_name = parts[1]
+    try:
+        revision = parts[3]
+    except IndexError:
+        raise ValueError('Revision cannot be extracted from url: {}'.format(url))
+
+    end_index = url.index(repo_name) + len(repo_name)
+    repo_url = url[:end_index]
+
+    return _strip_trailing_dot_git(repo_url), revision
+
+
+def _strip_trailing_dot_git(url):
+    if url.endswith('.git'):
+        url = url[:-len('.git')]
+    return url
+
+
+def is_github_repo_owner_the_official_one(context, repo_owner):
+    """Given a repo_owner, check if it matches the one configured to be the official one.
+
+    Args:
+        context (scriptworker.context.Context): the scriptworker context.
+        repo_owner (str): the repo_owner to verify
+
+    Raises:
+        scriptworker.exceptions.ConfigError: when no official owner was defined
+
+    Returns:
+        bool: True when ``repo_owner`` matches the one configured to be the official one
+
+    """
+    official_repo_owner = context.config['official_github_repos_owner']
+    if not official_repo_owner:
+        raise ConfigError(
+            'This worker does not have a defined owner for official GitHub repositories. '
+            'Given "official_github_repos_owner": {}'.format(official_repo_owner)
+        )
+
+    return official_repo_owner == repo_owner
+
+
+def _is_git_full_hash(commit_hash):
+    return _GIT_FULL_HASH_PATTERN.match(commit_hash) is not None
+
+
+def _check_github_url_is_supported(url):
+    if not is_github_url(url):
+        raise ValueError('"{}" is not a supported GitHub URL!'.format(url))
```

Comment:
Nit that isn't relevant for this PR: I wonder if it would make sense to have earlier validation for things like this?
Currently, we don't need `if` checks to assert that our variables are strings/numbers/whatever deep within our functions because we perform earlier validation of the task, and that allows the functions that operate on the task to be so much easier to read (and write).

In the same way. it could be beneficial to have validation of github urls in advance? It _could_ make logic cleaner, since we could leverage expectations with a data type (e.g.: "my function parameter is `GithubUrl`, so it better be a Github URL") rather than a function call ("what does `_check_github_url_is_supported(...)` do? Oh, it can throw an error, gotcha").

-----

This is a potentially subjective note. Not that big of a concern :)

Reply:
If, as you suggest, we weren't using strings everywhere (particularly if we could lean on type-checking), we could get away without these checks.

However, given that this is security sensitive code, I'd rather err on the side of having more sanity checks, rather than less.

Reply:
FWIW, I'm not recommending removing these checks, but rather moving them to the entry-point of untrusted data and doing validation there.
To represent that the string has been validated as the specialized type (github url string), we could use a type at that point.

But yeah, I agree, I don't think that we should lessen our sanity checks :)

Reply:
Agreed on keeping this check here, for now.